runtime/caml/osdeps.h

```diff
@@ -164,38 +164,97 @@ extern unsigned short caml_win32_minor;
    re-encoded in UTF-16.  The encoding of [s] is assumed to be UTF-8 if
    [caml_windows_unicode_runtime_enabled] is non-zero **and** [s] is valid
    UTF-8, or the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc], so it should be free
-   using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-8.
 */
+CAMLalloc(caml_stat_free, 1) CAMLreturns_nonnull()
 CAMLextern wchar_t* caml_stat_strdup_to_utf16(const char *s);
 
 /* [caml_stat_strdup_noexc_of_utf16(s)] returns a null-terminated copy of [s],
    re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero or
    the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc_noexc], so
-   it should be freed using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc_noexc], so it should
+   be freed using [caml_stat_free].
 
-   If allocation fails, this returns NULL.
+   If allocation fails, this returns NULL. This function never raises any
+   exceptions when [s] is valid UTF-16 but may raise [Sys_error] if it is not.
 */
+CAMLalloc(caml_stat_free, 1)
 CAMLextern char* caml_stat_strdup_noexc_of_utf16(const wchar_t *s);
 
 /* [caml_stat_strdup_of_utf16(s)] returns a null-terminated copy of [s],
    re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero or
    the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc_noexc], so
-   it should be freed using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
 
-   If allocation fails, this raises Out_of_memory.
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-16.
 */
+CAMLalloc(caml_stat_free, 1) CAMLreturns_nonnull()
 CAMLextern char* caml_stat_strdup_of_utf16(const wchar_t *s);
 
+/* [caml_stat_char_array_to_utf16(s, size, &out_size)] returns a copy of the
+   first [size] bytes of [s] re-encoded in UTF-16. The encoding of [s] is
+   assumed to be UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero
+   **and** [s] is valid UTF-8, or the current Windows code page otherwise. If
+   [out_size] is not [NULL], then the number of UTF-16 code units in the result
+   is recorded in [*out_size].
+
+   [size] must be greater than zero.
+
+   The returned buffer is allocated with [caml_stat_alloc], so it should be
+   freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory. This function may also raise
+   [Sys_error] if [s] is not valid UTF-8.
+*/
+CAMLmalloc(caml_stat_free, 1, 2) CAMLreturns_nonnull()
+CAMLextern wchar_t *caml_stat_char_array_to_utf16(const char *s, size_t size,
```

Comment:
The documentation comments do not explicitly mention that those functions are meant for `char*` data that may contain NUL characters. I think that it would be nice to point that out for people who may wonder when to use `caml_stat_char_array_to_os` instead of `caml_stat_strdup_to_os`.

Reply:
Agreed - I pushed an updated version